### ISSUE TITLE
chatlist stucks significantly while wiping up and down

### ISSUE THREAD

**Issue by csb0730** [deltachat/deltachat-android#1602]
- Delta Chat Version: 1.2.1, Android 4.1

Meanwhile I have around 320 chats and > 18000 messages in database.

When wiping the chatlist up and down it stucks significantly. This is a bad user experience meanwhile.
I suspect that this is caused by the big number of chats and/or the size of the database.

Is there any chance to get it flowing and smooth again?

**Comment by r10s** [comment]
> it stucks significantly

what does this mean in seconds? 10 seconds? 0.5 seconds? and how often and when does it appear? does it get better over time?

however, i do not thing there is a general way for improving that.

i personally have much more chats/messages and things are fine on a quite old moto g3, however, with a more recent android, but i suspect, that this is the issue.

you can try to profile, why the chatlist is so slow for you, eg. logging the time before/after the call to the core to see if it is on android- or on rust-side.

**Comment by gerryfrancis** [comment]
@csb0730 Do the lags occur in the recent version 1.12.4, too?

**Comment by csb0730** [comment]
To give you an impression:
When DC is opened it is more worse. If You wipe immediately down it stucks for 2-3s then it moves for some chats, then it stucks for 2s then it moves on ... If You want to move far below it hangs all few chats ... for approx 2s.
It seems as there is much activity in background and the list is loading some content of the chats which are coming in view.

When it is open and you move slowly it stucks only very short for 0.5s or so.

I mean, there is the first sentence of last message of each chat which is shown in chatlist entries. Maybe this is the reason for the behaviour? When is this loaded?

@r10s can you give me a hint what happens between the UI and the core while moving the chatlist? Are the chatlist's entries which will come into the view loaded dynamically?

**Comment by csb0730** [comment]
I did not try 12.4 (for information)

**Comment by r10s** [comment]
> can you give me a hint what happens between the UI and the core while moving the chatlist? Are the chatlist's entries which will come into the view loaded dynamically?

yip, like that. the chatlist is a so called "virtual list", items are created as needed. what might be, however, is that the issue is unreleated to the chatlist and that there is something blocking, for whatever reason. in theory, core should return immediately, however, in practice, we had issues there from time to time. but it was getting better with the last updates :)

**Comment by csb0730** [comment]
Ok more details:

When moving carefully up or down then there is a short stop with every new list entry which comes into the view. Regardless if moving up or down. Maybe this is a hint?

Where is the code point when a new entry gets in view? Is then the name of the chat and its last message load dynamically?

**Comment by r10s** [comment]
> When moving carefully up or down then there is a short stop with every new list entry which comes into the view

if it is every list entry - can you reproduce that also with a different account? maybe sth. is wrong with your database.

> Where is the code point when a new entry gets in view? Is then the name of the chat and its last message loaded dynamically?

should be `ConversationListAdapter::onBindViewHolder()`.

**Comment by csb0730** [comment]
Yes, it is every list entry. I strongly point to the database too, but why?
I can imagine a similar case while developing uuencode! There were some big messages into database which slowed down some functions dramatically. Maybe this is a similar case?!
Let you know ...

**Comment by csb0730** [comment]
This is the code:

```
 public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int i) {
    ViewHolder holder = (ViewHolder)viewHolder;
    DcChat chat = dcContext.getChat(dcChatlist.getChatId(i));
    DcLot summary = dcChatlist.getSummary(i, chat);
    holder.getItem().bind(dcContext.getThreadRecord(summary, chat), dcChatlist.getMsgId(i), summary, glideRequests, locale, batchSet, batchMode);
  }
```
I think the expensive operation is `dcChatlist.getSummary(i, chat)`?

**Comment by r10s** [comment]
this, and getChat() one line before accesses the database. typically this is fast, however.

**Comment by csb0730** [comment]
I think getChat() should be fast as this is retrieving only a small structure by primary key of the chat's table. However there is no index on chat_id.

But getSummary() which is done by Chatlist::get_summary() I think is the more expensive function. It loads the chat and a message from db.

Added some time measurement code into both functions and will check.

```
let t1 = Instant::now();
    code to measure ...
info!(context, "chatlist - get_summary(): {} ms", t1.elapsed().as_millis());
```
I hope this will show the problem.

**Comment by csb0730** [comment]
But one word to the performance in general:
With the switch from c-core to core-rust 0.500 -> 1.2.1 IMHO it felt like the performance was only half than before.
Whole app feels significantly slower now. Maybe you don't see this on a modern device but at my old phone it's obviously clear.

**Comment by csb0730** [comment]
This is an excerpt from performance log:

08-28 00:48:03.002  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 2 ms
08-28 00:48:03.642  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms
08-28 00:48:03.642  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:04.353  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  2 ms
08-28 00:48:04.353  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 0 ms
08-28 00:48:05.074  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  0 ms
08-28 00:48:05.074  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:06.015  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms
08-28 00:48:06.015  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:07.977  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms
08-28 00:48:07.977  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:09.058  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms
08-28 00:48:09.058  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 2 ms
08-28 00:48:10.139  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  2 ms
08-28 00:48:10.139  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:11.330  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms
08-28 00:48:11.330  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:12.651  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  3 ms
08-28 00:48:12.661  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:15.934  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  3 ms
08-28 00:48:15.934  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 0 ms
08-28 00:48:16.615  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms
08-28 00:48:16.615  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 2 ms
08-28 00:48:17.446  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  4 ms
08-28 00:48:17.446  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:18.076  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms
08-28 00:48:18.076  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:18.727  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  2 ms
08-28 00:48:18.727  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:19.528  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  0 ms
08-28 00:48:19.528  4167  4167 I DeltaChat: ThreadId(4) src/chatlist.rs:337: chatlist - get_summary(): 1 ms
08-28 00:48:19.698  4167  4167 I DeltaChat: ThreadId(4) src/chat.rs:496: chat - load_from_db():  1 ms

The timings seems to be ok? As there are only small load time for every entry this should not be the cause of the slow interface.
But on the other hand the DC now feels another bit slower than without time logging!

**Comment by r10s** [comment]
so, the timings seems to be okay, or?

> But on the other hand the DC now feels another bit slower than without time logging!

maybe it is just a bit slow device. tiny changes make the things directly recognizable slower as there is no more "power" left ;)

**Comment by csb0730** [comment]
I think loading the data from database itself seems not to be the constraint (regarding the timings measured).

But on the other hand it makes me suspicious that I can see and feel a significant slow down on chatlist behaviour when this time logging is enabled ! 2ms or even 10ms per chatlist entry cannot explain that.
You can move the list as slow as you want, but with every new upcoming list entry there is a short stop in list moving.

I (quite) don't see this with another install at same device. But this install only has a small db and only very few chats.

I think, to verify the behaviour I will create the opposite now: Empty the two functions so that they will not do the database query (and no log). => If then the movement is smoth we are back to the database ;-)

**Comment by csb0730** [comment]
Next result:
even when _chatlist->get_summary()_ and _chat->load_from_db()_ are empty and are doing nothing, moving the chatlist get stuck shortly after every new chat which is coming into the view!
Even when it's moved very slow!

What could it be now? What the list is doing while moving?

This is the list now without the support of _load_from_db()_ and _get_summary()_

![Screenshot_2020-08-28-20-55-38](https://user-images.githubusercontent.com/30560716/91606348-32708580-e972-11ea-8f9b-c5d7d2f377f6.png)

When the red marked lines are crossing the screen edge then the movement stucks shortly. This is reproducable.

Issue in java code?

**Comment by csb0730** [comment]
Maybe loading the avatar pictures?

**Comment by r10s** [comment]
thanks for profiling! good to know that the bottleneck is not the core.

i'd now try to remove more things until it feels "fast enough", starting with the avatars is reasonable. might be that this "fast enough" just never happens if the app is low on memory and the garbage collection is executed frequently. however, might also be that there is something that is worth optimizing and fixes the issue :)

**Comment by csb0730** [comment]
I you have some news please let me know. I can double check it.

**Comment by r10s** [comment]
in general, i am not working on the issue currently, esp. as i cannot really reproduce it. i am just trying to give some suggestions when being asked :)

**Comment by csb0730** [comment]
Maybe a misunderstanding, but next question: where can I find loading avatar?
I'm still trying ...

**Comment by r10s** [comment]
avatars seems to be loaded by `GroupRecordContactPhoto` via `Recipient` class (i just did a grep, i am also not deep in that code :) (the code is too complicated, btw, i assume half of the classes can be streamlined, but that is for another day)

**Comment by csb0730** [comment]
Thanks for that hint. My approach now is to reduce the code step by step until I'm getting a smooth list movement.
Then we'll see ;-)

**Comment by csb0730** [comment]
When having all code in _onBindViewHolder()_ disabled. Then list moves smooth and quick. No stuck at all.
That's a first success :)

**Comment by csb0730** [comment]
1. When disabling only this one line in _onBindViewHolder()_
`holder.getItem().bind(dcContext.getThreadRecord(summary, chat), dcChatlist.getMsgId(i), summary, glideRequests, locale, batchSet, batchMode);`
List moves smooth and quick.

2. Calling the methods (_bind()_ disabled) getMsgId(), getSummary() and getMsgId() in _onBindViewHolder()_ doesn't show a negative behaviour. Still smooth.

3. Leaving bind() enabled but disabling code for loading avatar, loading chat or get_summary (core) let stuck list while moving.

4. _onBindViewHolder()_ is called for every entry which is coming into view. This means several times per second.

5. When _bind()_ is disabled, a _Log.i("xxx")_ in _onBindViewHolder()_ doesn't have a negative impact for moving.

Summary:
The code which is done by bind() is responsable for the move delays.
But what is the root cause? How to simplify or disabling code? That seems a little bit tricky now.

**Comment by csb0730** [comment]
Wouldn't it be possible to block _bind()_ function until list is stop moving? I mean there is much deeper code which is done but not all of them is really necessary while moving. Later yes, ok.

**Comment by csb0730** [comment]
Found the big brake in code :-)

See (1) below
When replacing _getFreshMsgCount(chatId)_ by a dummy line with fixed value - see (2) - the code
is quick and UI is moving smooth: What a difference!

ApplicationDcContext.java:
```
  public ThreadRecord getThreadRecord(DcLot summary, DcChat chat) { // adapted from ThreadDatabase.getCurrent()
    int chatId = chat.getId();

    String body = summary.getText1();
    if (!body.isEmpty()) {
      body += ": ";
    }
    body += summary.getText2();

    Recipient recipient = getRecipient(chat);
    long date = summary.getTimestamp();
    //int unreadCount = getFreshMsgCount(chatId); // (1) <== this is the brake !
    int unreadCount = 0;  // (2)
    boolean verified = chat.isVerified();


    return new ThreadRecord(context, body, recipient, date,
        unreadCount, chatId,
        chat.getVisibility(), verified, chat.isSendingLocations(), summary);
  }
```
The last issue is now to speed up the related query of related core function in _chat.rs_:

```
    pub fn get_fresh_msg_cnt(self, context: &Context) -> usize {
        context
            .sql
            .query_get_value::<_, i32>(
                context,
                "SELECT COUNT(*)
               FROM msgs
              WHERE state=10
                AND hidden=0        <=== on 'hidden' column is no index ?! Maybe the root cause??
                AND chat_id=?;",
                params![self],
            )
            .unwrap_or_default() as usize
    }

```
Missing index on _hidden_ column in db? Maybe that's it?

**Comment by csb0730** [comment]
By missing an index, the number of messages counts!

**Comment by r10s** [comment]
good catch, @csb0730! thanks a lot! so get_fresh_msg_cnt seems to be the bottleneck - but i am wondering why.

> By missing an index, the number of messages counts!

there is an index over `chat_id` and over `state` - so the missing  index on `hidden` should not make much difference. the counting afterwards is probably `O(n)` (linear) anyway with or without index on `hidden`.

but this is just theory :)

more practically i am wondering if the scrolling is slow on all rows coming in view? or only when chats with many message come to view?

**Comment by csb0730** [comment]
I don't saw a difference while testing. Every new chat which comes in view shows the short stop when _getFreshMsgCount()_ is active. I'm wondering too, why all the remaining code doesn't drag the code, but this simple function! - at first view ;-)

Next step I'll verify how this query behaves in detail and then try to improve it. Maybe by an appropriate index. I'm not sure if this really mitigates query time.
My experience regarding database querys points in the direction that a missing index has terrible consequences. On the other hand an index really needs to match to the query to get successful.
We'll see.

The best success for now is, that it's found now :-)

By the way: I'm quite sure, that this issue is the root cause to have so much battery drain when working with DC!

**Comment by csb0730** [comment]
Ok, results:
I'm testing with a backup db (dbversion=65, 24000 msgs, size 121MB) at my PC with sqlitebrowser.

query is always the same (from _get_fresh_msg_count()_)
```
 SELECT COUNT(*)
               FROM msgs
              WHERE state=10
                AND hidden=0
                AND chat_id=20   <== several id's tested
```

1) original db unchanged:
query time => ~110 ms

2) added a single index for _hidden_ column:
`CREATE INDEX msgs_index7 ON msgs (hidden)`
query time => ~170 ms !!!

3) drop msgs_index7 and added multi column index:
`CREATE INDEX msgs_index8 ON msgs (state, hidden, chat_id)`
query time => **1-2 ms** !!!

I think it's clear now what has to be done :-)

**Comment by csb0730** [comment]
I'm just testing the first build. => I'm badly impressed by this progress!!!
Never saw DC so responsive :-)))

The chatlist is flying and the whole app is much quicker when moving around the chats  :-)

**Comment by csb0730** [comment]
What did I do at the end:

This is the key section from _open_ function in _sql.rs_
```
        if dbversion < 63 {
            info!(context, "[migration] v63");
            sql.execute("UPDATE chats SET grpid='' WHERE type=100", NO_PARAMS)?;
            sql.set_raw_config_int(context, "dbversion", 63)?;
        }
        // cs: this is the fix
        info!(context, "[fix for #1602] speed up get_fresh_msg_count()");
        sql.execute("CREATE INDEX if not exists msgs_multi on msgs (state, hidden, chat_id)", NO_PARAMS)?;
```
See the CREATE INDEX line above. => That's all !

:-)

**Comment by csb0730** [comment]
@r10s Finally this issue belongs to the core. Maybe You can move it?

Would you add the key lines to the code? I do not have current master to do a PR.

We should not wait with this fix ;-)

**Comment by r10s** [comment]
i move the issue, thanks a lot for profiling!
